llvm/lib/Target/Mips/NanoMipsCompressJumpTables.cpp

```diff
@@ -0,0 +1,163 @@
+#include "llvm/CodeGen/MachineBasicBlock.h"
+#include "llvm/CodeGen/MachineFunction.h"
+#include "llvm/CodeGen/MachineFunctionPass.h"
+#include "llvm/CodeGen/MachineJumpTableInfo.h"
+
+#include <cmath>
+
+using namespace llvm;
+
+#define NM_COMPRESS_JUMP_TABLES_OPT_NAME                                       \
+  "nanoMIPS compress jump tables optimization pass"
+
+namespace {
+struct NMCompressJumpTables : public MachineFunctionPass {
+  static char ID;
+  const MipsSubtarget *STI;
+  const TargetInstrInfo *TII;
+  MachineFunction *MF;
+  SmallVector<int, 8> BlockInfo;
+  SmallVector<int, 8> BrOffsets;
+
+  int computeBlockSize(MachineBasicBlock &MBB);
+  void scanFunction();
+  bool compressJumpTable(MachineInstr &MI, int Offset);
+
+  NMCompressJumpTables() : MachineFunctionPass(ID) {}
+  StringRef getPassName() const override {
+    return NM_COMPRESS_JUMP_TABLES_OPT_NAME;
+  }
+  bool runOnMachineFunction(MachineFunction &Fn) override;
+};
+} // namespace
+
+char NMCompressJumpTables::ID = 0;
+
+int NMCompressJumpTables::computeBlockSize(MachineBasicBlock &MBB) {
+  // return MBB.size() * 6;
+  int Size = 0;
+  for (const MachineInstr &MI : MBB)
+    Size += TII->getInstSizeInBytes(MI);
+  return Size;
+}
+
+void NMCompressJumpTables::scanFunction() {
+  BlockInfo.clear();
+  BlockInfo.resize(MF->getNumBlockIDs());
+  BrOffsets.clear();
+  bool findBR = MF->getJumpTableInfo() &&
+                !MF->getJumpTableInfo()->getJumpTables().empty();
+  if (findBR)
+    BrOffsets.resize(MF->getJumpTableInfo()->getJumpTables().size());
+  int Offset = 0;
+  for (MachineBasicBlock &MBB : *MF) {
+    BlockInfo[MBB.getNumber()] = Offset;
+    Offset += computeBlockSize(MBB);
+    if (findBR)
+      for (auto &MI : MBB) {
+        if (MI.getOpcode() == Mips::BRSC_NM) {
+          int JTIdx = MI.getOperand(1).getIndex();
+          BrOffsets[JTIdx] = Offset;
+          break;
+        }
+      }
+  }
+}
+
+bool NMCompressJumpTables::compressJumpTable(MachineInstr &MI, int Offset) {
+  if (MI.getOpcode() != Mips::JumpTableDest32NM)
+    return false;
+
+  int JTIdx = MI.getOperand(3).getIndex();
+  auto &JTInfo = *MF->getJumpTableInfo();
+  const MachineJumpTableEntry &JT = JTInfo.getJumpTables()[JTIdx];
+
+  // The jump-table might have been optimized away.
+  if (JT.MBBs.empty())
+    return false;
+
+  int MaxOffset = std::numeric_limits<int>::min(),
+      MinOffset = std::numeric_limits<int>::max();
+  int BrOffset = BrOffsets[JTIdx];
+
+  bool Signed = false;
+  for (auto Block : JT.MBBs) {
+    int BlockOffset = BlockInfo[Block->getNumber()];
+    MaxOffset = std::max(MaxOffset, BlockOffset - BrOffset);
+    MinOffset = std::min(MinOffset, BlockOffset - BrOffset);
+    if (MinOffset < 0)
+      Signed = true;
+  }
+
+  if (std::max(std::abs(MinOffset), MaxOffset) == MinOffset)
+    MaxOffset = MinOffset;
+
+  auto MFI = MF->getInfo<MipsFunctionInfo>();
+  MCSymbol *JTS = MFI->getJumpTableSymbol(JTIdx);
+
+  if (Signed && isInt<8>(MaxOffset)) {
```

Comment:
This can be mostly folded away if the JumpTableDest instructions are unified:
```c++
if (isInt<8>(MaxOffset))
  MFI->setJumpTableEntryInfo(JTIdx, 1, JTS, Signed);
else if (isInt<16>(MaxOffset))
  MFI->setJumpTableEntryInfo(JTIdx, 2, JTS, Signed);

```
?